File: test/harness/kysely/smoke.ts
// Pins semantic invariants in the type system so `tsc --strict` catches a
// regression that would otherwise type-check: Generated<T> must unwrap to T on
// select and go optional on insert, a database-side default must be the only
// thing that becomes Generated<T>, a DateTime column must select as Date, an
// optional column must be `T | null`, an enum column must stay a value union
// of the @map-ped database values, the DB interface must be keyed by the
// @@map-ped table names, and the implicit m2m join tables must exist.
import {
  DummyDriver,
  type Insertable,
  Kysely,
  PostgresAdapter,
  PostgresIntrospector,
  PostgresQueryCompiler,
  type Selectable,
  type Updateable,
} from 'kysely'

import type * as t from './types'

const db = new Kysely<t.DB>({
  dialect: {
    createAdapter: () => new PostgresAdapter(),
    createDriver: () => new DummyDriver(),
    createIntrospector: (database) => new PostgresIntrospector(database),
    createQueryCompiler: () => new PostgresQueryCompiler(),
  },
})

export const accountQuery = db
  .selectFrom('accounts')
  .select(['id', 'created_at', 'status'])
  .compile()

export const joinQuery = db
  .selectFrom('_PostToTag')
  .innerJoin('Post', 'Post.id', '_PostToTag.A')
  .innerJoin('Tag', 'Tag.id', '_PostToTag.B')
  .select(['Post.title', 'Tag.label'])
  .compile()

export const castQuery = db.selectFrom('_cast').select(['A', 'B']).compile()

type Account = Selectable<t.Account>
type Profile = Selectable<t.Profile>
type Board = Selectable<t.Board>
type Sequence = Selectable<t.Sequence>
type Torture = Selectable<t.Torture>
type Inventory = Selectable<t.Inventory>

type Expect<T extends true> = T
type Equal<A, B> =
  (<T>() => T extends A ? 1 : 2) extends <T>() => T extends B ? 1 : 2 ? true : false

export type Cases = [
  Expect<Equal<Account['tags'], string[]>>,
  Expect<Equal<Account['bigNum'], bigint>>,
  Expect<Equal<Account['status'], 'ACTIVE' | 'INACTIVE' | 'PENDING_REVIEW'>>,
  Expect<Equal<Account['created_at'], Date>>,
  Expect<Equal<Account['raw'], Buffer>>,
  Expect<Equal<Profile['bio'], string | null>>,
  Expect<Equal<Profile['age'], number | null>>,
  Expect<Equal<Board['visibility'], 'public' | 'private' | 'link_only'>>,
  Expect<Equal<Board['fallback'], 'public' | 'private' | 'link_only' | null>>,
  Expect<Equal<Board['audiences'], ('public' | 'private' | 'link_only')[]>>,
  Expect<Equal<Sequence['id'], bigint>>,
  Expect<Equal<Torture['born'], Date>>,
  Expect<Equal<Inventory['codes'], number[]>>,

  // autoincrement() and dbgenerated() are evaluated by the database, so the
  // column carries a DDL default and the insert may omit it.
  Expect<Equal<Insertable<t.Sequence>['id'], bigint | undefined>>,
  Expect<Equal<Insertable<t.Computed>['id'], string | undefined>>,
  // uuid()/uuid(7)/ulid()/cuid(2)/nanoid(10) are evaluated by prisma's client,
  // which leaves the column without a DDL default — a raw kysely insert has to
  // provide the value itself.
  Expect<Equal<Insertable<t.Post>['id'], string>>,
  Expect<Equal<Insertable<t.Event>['id'], string>>,
  Expect<Equal<Insertable<t.Ticket>['id'], string>>,
  Expect<Equal<Insertable<t.Badge>['id'], string>>,
  Expect<Equal<Insertable<t.Coupon>['id'], string>>,

  // The generated Generated<T> must preserve the insert/update sides of a
  // nested ColumnType, not collapse them back to the select type.
  Expect<Equal<Insertable<t.Account>['created_at'], Date | string | undefined>>,
  // Updateable makes every key optional, so undefined joins the update type.
  Expect<Equal<Updateable<t.Account>['created_at'], Date | string | undefined>>,
  // @updatedAt without @default is Prisma-Client-managed, not DB-managed, so
  // a raw kysely insert must still provide it.
  Expect<Equal<Insertable<t.Profile>['updated_at'], Date | string>>,
]

File: test/harness/django/smoke.py
"""Pins semantic invariants in the type system so mypy --strict catches a
regression that would otherwise type-check: an optional scalar must stay
`| None`, a self-relation must resolve to the model class, and an enum column
must stay a str-valued choices field.

Run as a script it also exercises the parts mypy cannot see: the system-check
framework (field-name rules, choices, composite keys, related_name clashes),
migration serialization (a lambda default or an unserializable db_default only
fails here), offline PostgreSQL DDL compilation, model instantiation, which
fires every client-side default (uuid4/uuid7/ulid string generators), and
query compilation, which is the only check that resolves the joins."""

import os

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "settings")

import django

django.setup()

import uuid

from app.models import (
    Account,
    Board,
    Category,
    Claim,
    Event,
    Inventory,
    Monarch,
    Post,
    Profile,
    Tag,
    Ticket,
    Torture,
)


def _static_smoke(account: Account, profile: Profile) -> None:
    bio: str | None = profile.bio
    age: int | None = profile.age
    parent: Category | None = Category().parent
    successor: Monarch | None = Monarch().successor
    status: str = account.status
    _ = (bio, age, parent, successor, status)


def _runtime_smoke() -> None:
    from django.apps import apps
    from django.core.checks import run_checks
    from django.db import connection
    from django.db.migrations.autodetector import MigrationAutodetector
    from django.db.migrations.loader import MigrationLoader
    from django.db.migrations.state import ProjectState
    from django.db.migrations.writer import MigrationWriter

    # Warnings count too: the ones that apply here (a unique ForeignKey that
    # should be a OneToOneField, a nullable ManyToManyField, a symmetrical
    # self-relation that silently drops related_name) each mean the generated
    # model says something other than the Prisma schema does.
    assert run_checks() == []

    # Rendering the autodetected initial migration runs the migration
    # serializer over every field: this is what rejects a lambda default or an
    # expression without deconstruct() support. The app needs a migrations
    # package for the autodetector to consider it at all — without one this
    # loop is empty and proves nothing, so the assertions guard that first.
    loader = MigrationLoader(None, ignore_no_migrations=True)
    autodetector = MigrationAutodetector(loader.project_state(), ProjectState.from_apps(apps))
    changes = autodetector.changes(graph=loader.graph)
    assert "app" in changes, "autodetector saw no models: is app/migrations/__init__.py missing?"
    for migration in changes["app"]:
        assert len(MigrationWriter(migration).as_string()) > 0

    # collect_sql compiles the full PostgreSQL DDL without a live server.
    with connection.schema_editor(collect_sql=True, atomic=False) as editor:
        for model in apps.get_app_config("app").get_models():
            editor.create_model(model)
    assert len(editor.collected_sql) > 0

    # Instantiation applies every client-side default.
    for model in apps.get_app_config("app").get_models():
        model()

    uuid.UUID(Account().id)
    assert uuid.UUID(Event().id).version == 7
    assert len(Ticket().id) == 26


def _default_value_smoke() -> None:
    """Instantiation applies every `default=`, so this reads back the value the
    generated source only spells out. The golden masters pin the text of the
    default; what it evaluates to in Python is checked here."""

    # An enum default carries the mapped database value, not the Prisma name.
    assert Board().visibility == "link_only"

    torture = Torture()
    assert torture.quoted == 'it\'s a "quote" and a \\ backslash'
    assert torture.unicode == "日本語と🔥絵文字"
    assert torture.empty == ""
    # Beyond Number.MAX_SAFE_INTEGER: the literal must survive as an int.
    assert torture.big_pos == 9007199254740993
    assert torture.big_neg == -9007199254740993
    assert torture.json_obj == {"a": 1, "b": [True, None, "x"]}
    assert torture.json_arr == []
    assert torture.json_str == "quoted"
    # USE_TZ is on, so a naive default would warn and compare wrongly.
    assert torture.born.tzinfo is not None

    inventory = Inventory()
    assert inventory.tags == []
    assert inventory.codes == [1, 2, 3]
    assert inventory.labels == ["a", "b"]

    # The point of emitting a function rather than a literal: two instances must
    # not share one mutable default.
    first = Inventory()
    first.codes.append(99)
    assert Inventory().codes == [1, 2, 3]
    assert Torture().json_obj is not Torture().json_obj


def _query_smoke() -> None:
    """Compiling a QuerySet to SQL resolves the relations end to end, which
    neither the system checks nor the DDL reach: a related_name pointing at the
    wrong field, a through model whose columns do not match the join table, or a
    to_field aimed at the primary key instead of the referenced unique column
    all produce a wrong join here while passing everything above."""

    # Forward and reverse traversal across a ForeignKey and a OneToOneField.
    assert '"post"."author_id" = "accounts"."id"' in str(
        Post.objects.select_related("author").filter(author__status="ACTIVE").query
    )
    assert '"profile"."account_id"' in str(Account.objects.filter(profile__verified=True).query)

    # The many-to-many joins through Prisma's own table and its "A"/"B" columns,
    # in both directions.
    post_to_tag = str(Post.objects.filter(tags__label="a").query)
    assert 'INNER JOIN "_PostToTag" ON ("post"."id" = "_PostToTag"."A")' in post_to_tag
    assert 'INNER JOIN "tag" ON ("_PostToTag"."B" = "tag"."id")' in post_to_tag
    tag_to_post = str(Tag.objects.filter(posts__title="x").query)
    assert 'INNER JOIN "_PostToTag" ON ("tag"."id" = "_PostToTag"."B")' in tag_to_post

    # Two relations to the same model must not collapse into one join.
    followers = str(Account.objects.filter(followers__follower__id="1").query)
    assert '"follow"."following_id"' in followers
    assert '"follow"."follower_id"' in followers

    # Self-relations join the table to an alias of itself.
    assert 'INNER JOIN "category" T2' in str(Category.objects.filter(children__name="x").query)
    assert 'INNER JOIN "monarch" T2' in str(Monarch.objects.filter(predecessor__name="x").query)

    # A foreign key aimed at a unique column that is not the primary key.
    assert 'INNER JOIN "handle" ON ("claim"."slug" = "handle"."slug")' in str(
        Claim.objects.select_related("handle").query
    )


if __name__ == "__main__":
    _runtime_smoke()
    _default_value_smoke()
    _query_smoke()
    print("django smoke OK")
